mdpow/ensemble.py

```diff
@@ -0,0 +1,428 @@
+and has methods analogous the the Universe object. The main one being select_atoms which returns
+a EnsembleAtomGroup. An Ensemble in its current form can also be built by manually adding and popping
+ universes into an empty instance.
+
+
+EnsembleAtomGroup:
+-----------------
+
+The EnsembleAtomGroup is created by running the select_atoms method on an Ensemble. It stores AtomGroup
+selections of the groups generated by running select atom on individual universes in a dictionary with
+the same key structure as the parent Ensemble class.It returns a copy of the parent Ensemble object when
+ the ensemble method is run.
+
+EnsembleAnalysis:
+-----------------
+
+The EnsembleAnalysis is a class inspired by the AnalysisBase in MDAnalysis which iterates over the systems
+in the ensemble and the frames in the systems. It sets up both iterations between universes and universe
+frames allowing for analysis to be run on both whole systems and the frames of those systems. This allows
+for users to easily run analyses on MDPOW simulations.
+"""
+
+import os
+
+from gromacs.utilities import in_dir
+import MDAnalysis as mda
+from MDAnalysis.lib.log import ProgressBar
+from MDAnalysis.exceptions import FileFormatWarning, NoDataError, MissingDataWarning, SelectionError
+import numpy as np
+
+import logging
+
+logger = logging.getLogger('mdpow.ensemble')
+
+
+class NoDataWarning(Exception):
+    pass
+
+
+class Ensemble(object):
+    """Collection of related MDAnalysis Universe objects
+
+    Stores systems produced by running mdpow-fep organized
+    by solvent, interaction, and lambda.
+
+    Given a mdpow simulation directory will load the MD
+    simulation files with the directory structure as keys.
+
+    Typical workflow for MDPOW directory::
+
+        ens = Ensemble(dirname='molecule')
+
+    Typical workflow for manual addition::
+
+        ens = Ensemble()
+        ens.add_system(topology='md.gro', trajectory='md.xvg')
+
+    """
+
+    def __init__(self, dirname=None, solvents=['octanol', 'water'], **kwargs):
+        """
+
+        """
+        try:
+            self.top: dict = kwargs.pop('top_paths')
+        except KeyError:
+            self.top = None
+
+        self.num_systems = 0
+        self.ensemble = {}
+        self.system_keys = []
+        self.solvents = solvents
+
+        if dirname is None:
+            return
+
+        if not os.path.exists(dirname):
+            logger.error(f"Directory {dirname} does not exist")
+        self.ensemble_dir = dirname
+        self.build_ensemble()
+
+    def __repr__(self):
+        return f"<Ensemble Containing {self.num_systems} System>"
+
+    def __len__(self):
+        return self.num_systems
+
+    def __getitem__(self, index):
+        """Allows dictionary like indexing"""
+        try:
+            item = self.ensemble[index]
+            return item
+        except KeyError:
+            logger.error(KeyError("Invalid system key"))
+
+    def __iter__(self):
```

Comment:
no I'll add one

Reply:
I decided to just remove the iter method since its not used, and not needed in EnsembleAnalysis, also makes more sense to iterate by using the list of keys to record which system in being iterated on.